### PR TITLE
[MoveosStd] Refactor singleton and display

### PR DESCRIPTION
**Pull request by jolestar**
## Summary

Part of #1066 

Refactor singleton and display.

1. Make singleton and display protected by `private_generics<T>`.
2. Make the display object default permanent and support Display<Object<T>>
3. Rename `context::borrow_object_mut` to `context::borrow_mut_object`.
4. Add more tests to the object.

### Refactor Display

Currently, Display has an issue:

We can create an `Object<Display<T>>` to display the `T`, But if we want to display the `Object<T>`, the display template can not get the `ObjectID` and `owner`.

So the Display object should be `Object<Display<Object<T>>>`, and the template can be:

```move
struct NFT has key,store {
        name: String,
        collection: ObjectID,
        creator: address,
}
```

```move
let display_obj: Object<Display<Object<T>>> = display::object_display<NFT>(ctx);
display::set_value(display_obj, string::utf8(b"url"), string::utf8(b"https://baseurl/{value.collection}/{id}")) 
```

So, I provide `display::resource_display<T>` for `T`, and `display::object_display<T>` for `Object<T>`.

- Closes #issue

### CODE REVIEW

File: moveos/moveos-stdlib/moveos-stdlib/sources/context.move [hunk 1]
```diff
@@ -94,18 +94,44 @@ module moveos_std::context {
...
+    public fun borrow_singleton<T: key>(self: &Context): &Object<T> {
+        let object_id = object::singleton_object_id<T>();
+        borrow_object(self, object_id)
+    }
+
     /// Borrow mut Object from object store with object_id
-    /// If the object is not shared, only the owner can borrow an &mut Object from the global object storage
-    public fun borrow_object_mut<T: key>(_self: &mut Context, owner: &signer, object_id: ObjectID): &mut Object<T> {
+    /// If the object is not shared, only the owner can borrow an `&mut Object<T>` from the global object storage
+    public fun borrow_mut_object<T: key>(_self: &mut Context, owner: &signer, object_id: ObjectID): &mut Object<T> {
```
Comment (baichuan3) [inline moveos/moveos-stdlib/moveos-stdlib/sources/context.move:134]: Since the `Context` reference is not used, can it be removed from the parameter?

Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/context.move:134]: We require the `&mut Context` for ensure that if a transaction change the state, it must include `&mut Context` in the entry function's arguments.

File: moveos/moveos-stdlib/moveos-stdlib/sources/event.move [hunk 2]
```diff
@@ -47,7 +47,7 @@ module moveos_std::event {
     /// Borrow a mut event handle from the object storage
     fun borrow_event_handle_mut<T>(ctx: &mut Context): &mut EventHandle {
         let event_handle_id = derive_event_handle_id<T>();
```
Comment (baichuan3) [inline moveos/moveos-stdlib/moveos-stdlib/sources/event.move:49]: `event_handle_id` can also be refactored to use singleton_object directly

File: moveos/moveos-stdlib/moveos-stdlib/sources/context.move (hunk 1 above)
Comment (baichuan3) [inline moveos/moveos-stdlib/moveos-stdlib/sources/context.move:134]: `borrow_xxx_mut` and `borrow_mut_xxx` can be unified in rooch framework

Comment (baichuan3) [reply moveos/moveos-stdlib/moveos-stdlib/sources/context.move:134]: Is this just a convention, or is it mandatory (the verifier checks)?

File: moveos/moveos-stdlib/moveos-stdlib/sources/event.move (hunk 2 above)
Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/event.move:49]: I plan to remove the EventHandle from state. #1056

File: moveos/moveos-stdlib/moveos-stdlib/sources/context.move (hunk 1 above)
Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/context.move:134]: It ensure by verifier and vm. An `&mut T` only can borrow from `&mut Context` or `&mut Object<T>` entry function argument. If the entry function has not `&mut` argument, it is an readonly function. I record an issue #1078

Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/context.move:134]: Only one `borrow_box_mut` in the raw_table is an private native function. We can refactor it in the future.

Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/context.move:134]: @stevenlaw123

**Review by baichuan3 — APPROVED** [review]

**Pull request merged**